src/Compilers/CSharp/Portable/BoundTree/BoundDagEvaluation.cs

```diff
@@ -31,11 +32,22 @@ public override int GetHashCode()
         }
         public static bool operator ==(BoundDagEvaluation left, BoundDagEvaluation right)
         {
-            return (left == (object)null) ? right == (object)null : left.Equals(right);
+            return (left is null) ? right is null : left.Equals(right);
         }
         public static bool operator !=(BoundDagEvaluation left, BoundDagEvaluation right)
         {
             return !(left == right);
         }
     }
+
+    partial class BoundDagIndexEvaluation
+    {
+        public override bool Equals(object obj) => obj is BoundDagIndexEvaluation other && this.Equals(other);
+        public override bool Equals(BoundDagEvaluation obj) => obj is BoundDagIndexEvaluation other && this.Equals(other);
+        public override int GetHashCode() => base.GetHashCode() ^ this.Index;
+        public bool Equals(BoundDagIndexEvaluation other)
```

Comment:
>public bool Equals(BoundDagIndexEvaluation other) [](start = 8, length = 49)

Are we calling this overload directly, other than in overloads above? If not, then it seems like we can get away with one overload in this class.
```
public override bool Equals(BoundDagEvaluation obj)
{
    return base.Equals(other) &&
        this.Index == ((BoundDagIndexEvaluation).Index;
}
```